doc/design/docker-cp.md

```diff
@@ -0,0 +1,144 @@
+
+The personality code should be fairly straight forward, the goal is to call the portlayer for the copy while passing the tar stream either to the portlayer or from it. The best way to do this is to hook the writiers together using a PipeReader and then shepherding the data in the direction it needs to go. 
+
+## Portlayer Design
+
+ There will be two situations to be concerned with in the portlayer. When the container is *on* and when the container is *off*.
+ 
+ 
+### Off Container Scenario
+ 
+ This is the case where guest tools is not needed since the guest in off. This means that all action will need to occur on the VCH. There are several concerns and many of them are listed in this ticket [here](https://github.com/vmware/vic/issues/4105)
+ 
+ The most concern surrounding this piece of work will be surrounding VCH restart and confirming that the disk is still left in a good state if a copy was occuring mid restart. Additionally we need to block a container from starting when we have the r/w layer mounted. While that container is blocked from being started we might look at the possibility of returning a 500 error for a start invocation on a blocked container. We must also block other copy operations until the current one is successful, one major reason is because we cannot do any kind of sanity checking on available space with concurrent writes to the same disk. 
+ 
+ Additionally, it must be ensured that regardless of error or success the disk must be detached from the VCH. We must ensure that this also occurs when we boot up the VCH in the event that the disk is attached to the VCH and is hard shutdown. If the VCH does not already have a disk detach functionality at boot time we should be adding this. This is an ideal time to invoke calls to `fsck` and `journal` in order to recover from mid copy shutdowns. Since the VCH is meant to be stateless(and the portlayer) we should aim to jettison and disk that is attached to the VCH at boot time if we do write this portion of the portlayer(We should even make it a separate ticket as testing and ensuring quality on this could take time.)
+ 
+ In An Ideal scenario without any failures the process will be as such : 
+ 
+ ```
+ 1) get request from personality for a file copy, it will have the stream to readfrom/write to as well as a path and a container ID.
+ 
+ 2) depending on the endpoint hit depends on the to/from scenario either way in an *off* scenario we would first get the container object and confirm that it is off. 
+ 
+ 3) Once confirmed that the container is off we will need to immediately block starts(and even wait if one is triggered between the status check and requesting the lockdown). In addition to this, we must block other `cp` operations until the current one is complete. 
+ 
+ 4) Then we will mount the r/w layer at a temp directory location. and assemble that path along with the target copy path.
```

Comment:
The above suggests that docker may remount vanilla volumes for the copy. I do not think this can be garunteed for their plugin based behavior though. I will take a look through their code.